Gems/ROS2/Code/Source/Imu/ROS2ImuSensorComponent.cpp

```diff
@@ -77,25 +77,61 @@ namespace ROS2
+        const auto fullTopic = ROS2Names::GetNamespacedName(GetNamespace(), publisherConfig.m_topic);
+        m_imuPublisher = ros2Node->create_publisher<sensor_msgs::msg::Imu>(fullTopic.data(), publisherConfig.GetQoS());
+
+        m_linearAccelerationCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_linearAccelerationVariance);
+        m_angularVelocityCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_angularVelocityVariance);
+        m_orientationCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_orientationVariance);
+
+        StartSensor(
+            m_sensorConfiguration.m_frequency,
+            [this](float imuDeltaTime, AzPhysics::SceneHandle sceneHandle, float physicsDeltaTime)
+            {
+                if (!m_sensorConfiguration.m_publishingEnabled)
+                {
+                    return;
+                }
+                OnImuEvent(imuDeltaTime, sceneHandle, physicsDeltaTime);
+            },
+            [this]([[maybe_unused]] AzPhysics::SceneHandle sceneHandle, float physicsDeltaTime)
+            {
+                OnPhysicsEvent(sceneHandle);
+            });
     }
 
-    void ROS2ImuSensorComponent::OnPhysicsInitialization(AzPhysics::SceneHandle sceneHandle)
+    void ROS2ImuSensorComponent::Deactivate()
     {
-        AzPhysics::RigidBody* rigidBody = nullptr;
-        AZ::EntityId entityId = GetEntityId();
-        Physics::RigidBodyRequestBus::EventResult(rigidBody, entityId, &Physics::RigidBodyRequests::GetRigidBody);
-        AZ_Assert(rigidBody, "Entity %s does not have rigid body.", entityId.ToString().c_str());
-
-        m_bodyHandle = rigidBody->m_bodyHandle;
+        StopSensor();
+        m_bodyHandle = AzPhysics::InvalidSimulatedBodyHandle;
+        m_imuPublisher.reset();
     }
 
-    void ROS2ImuSensorComponent::OnPhysicsSimulationFinished(AzPhysics::SceneHandle sceneHandle, float deltaTime)
+    void ROS2ImuSensorComponent::OnPhysicsEvent(AzPhysics::SceneHandle sceneHandle)
     {
+        if (m_bodyHandle == AzPhysics::InvalidSimulatedBodyHandle)
+        {
+            AzPhysics::RigidBody* rigidBody = nullptr;
+            AZ::EntityId entityId = GetEntityId();
+            Physics::RigidBodyRequestBus::EventResult(rigidBody, entityId, &Physics::RigidBodyRequests::GetRigidBody);
+
+            if (!rigidBody)
+            {
+                AZ_Error("ROS2ImuSensorComponent", false, "Entity %s does not have a rigid body - stopping Imu sensor.", entityId.ToString().c_str());
+                StopSensor();
+                return;
+            }
+
+            m_bodyHandle = rigidBody->m_bodyHandle;
+        }
+
         auto* sceneInterface = AZ::Interface<AzPhysics::SceneInterface>::Get();
-        const auto gravity = sceneInterface->GetGravity(sceneHandle);
         auto* body = sceneInterface->GetSimulatedBodyFromHandle(sceneHandle, m_bodyHandle);
```

Comment:
I would suggest multiple assertions here - the code can mistakenly be executed in a different context by mistake (e.g. in Asset Processor) and a good assertion can be helpful.

Reply:
I belive gravity is not needed.

Reply:
I added an assert on body pointer - this is a good catch. I left sceneInterface as it is - do you maybe think that it is necessary to assert it, when it is a result from:
   AZ::Interface<AzPhysics::SceneInterface>::Get()
engine call? Here I assume, that it won't rather fail, and if it will, then we have some critical issue. Is such reasoning not correct?